### PR TITLE
Feature claim

### PR DESCRIPTION
**Pull request by AndriiAndrusyshyn**
## Consensus Changes
- [x] Consensus Changes

## API Changes
- [ ] API Changes

## Documentation Additions
- [ ] Documentation Additions

### CODE REVIEW

File: contracts/contracts/rem.system/include/rem.system/rem.system.hpp [hunk 1]
```diff
@@ -273,8 +274,8 @@ namespace eosiosystem {
        *  stated.amount * 2 ^ (weeks_since_launch/weeks_per_year) * (time_to_mature / mature_period)
        */
       double              last_vote_weight = 0; /// the vote weight cast the last time the vote was updated
-      time_point          vote_mature_time;
-
+      time_point          locked_stake_period;
+      time_point          last_claim_time;
```
Comment (me-vadim) [inline contracts/contracts/rem.system/include/rem.system/rem.system.hpp:279]: what is the purpose of adding this? I don`t see it beeing used

File: contracts/contracts/rem.system/include/rem.system/rem.system.hpp [hunk 2]
```diff
@@ -301,7 +302,7 @@ namespace eosiosystem {
       bool vote_is_reasserted() const;
 
       // explicit serialization macro is not necessary, used here only to improve compilation time
-      EOSLIB_SERIALIZE( voter_info, (owner)(proxy)(producers)(staked)(last_vote_weight)(vote_mature_time)(proxied_vote_weight)(is_proxy)(flags1)(reserved2)(reserved3)(last_reassertion_time) )
+      EOSLIB_SERIALIZE( voter_info, (owner)(proxy)(producers)(staked)(last_vote_weight)(locked_stake_period)(proxied_vote_weight)(is_proxy)(flags1)(reserved2)(reserved3)(last_reassertion_time) )
```
Comment (me-vadim) [inline contracts/contracts/rem.system/include/rem.system/rem.system.hpp]: last_claim_time hasn`t been mentioned in EOSLIB_SERIALIZE

File: contracts/contracts/rem.system/src/delegate_bandwidth.cpp [hunk 3]
```diff
@@ -249,17 +249,17 @@ namespace eosiosystem {
          voter_itr = _voters.emplace( voter, [&]( auto& v ) {
             v.owner            = voter;
             v.staked           = total_update.amount;
-            v.vote_mature_time = current_time_point() + vote_mature_period;
+            v.locked_stake_period = current_time_point() + vote_mature_period;
```
Comment (mark-tsyrulnyk) [inline contracts/contracts/rem.system/src/delegate_bandwidth.cpp]: @AndriiAndrusyshyn could you please rename `vote_mature_period` -> `stake_lock_period` to be consistent with new names

File: contracts/contracts/rem.system/src/delegate_bandwidth.cpp [hunk 4]
```diff
@@ -249,17 +249,17 @@ namespace eosiosystem {
...
          });
       } else {
          _voters.modify( voter_itr, same_payer, [&]( auto& v ) {
             v.staked += total_update.amount;
             if ( total_update.amount >= 0 ) {
                const auto restake_rate = double(total_update.amount) / v.staked;
                const auto prevstake_rate = 1 - restake_rate;
-               const auto time_to_mature = std::max( v.vote_mature_time - current_time_point(), microseconds{} );
+               const auto time_to_mature = std::max( v.locked_stake_period - current_time_point(), microseconds{} );
```
Comment (mark-tsyrulnyk) [inline contracts/contracts/rem.system/src/delegate_bandwidth.cpp]: @AndriiAndrusyshyn same here, `time_to_stake_unlock` (or something like this)

File: contracts/contracts/rem.system/src/voting.cpp [hunk 5]
```diff
@@ -79,17 +79,28 @@ namespace eosiosystem {
...
+             p.locked_stake = tot.own_stake_amount;
+         });
       }
 
    }
 
    void system_contract::unregprod( const name& producer ) {
       require_auth( producer );
 
-      const auto& voter = _voters.get( producer.value, "producer is not found in voter table" );
```
Comment (mark-tsyrulnyk) [inline contracts/contracts/rem.system/src/voting.cpp:89]: @AndriiAndrusyshyn please keep this logic
```
const auto& voter = _voters.get( producer.value, "producer is not found in voter table" );
check( voter.vote_mature_time <= current_time_point(), "producers are not allowed to unregister during stake lock period" );

claimrewards(producer);
```

File: contracts/contracts/rem.system/src/voting.cpp [hunk 6]
```diff
@@ -79,17 +79,28 @@ namespace eosiosystem {
...
-      check( voter.vote_mature_time <= current_time_point(), "producers are not allowed to unregister during stake lock period" );
+      claimrewards(producer);
 
-      claimrewards( producer );
+      if(is_block_producer(producer)) {
```
Comment (mark-tsyrulnyk) [inline contracts/contracts/rem.system/src/voting.cpp]: @AndriiAndrusyshyn this check is redundant

File: contracts/contracts/rem.system/src/voting.cpp [hunk 7]
```diff
@@ -79,17 +79,28 @@ namespace eosiosystem {
...
+          const auto &vot = _voters.get(producer.value, "user has no resources");
+          _voters.modify(vot, producer, [&](auto &p) {
+              p.locked_stake_period = current_time_point() + stake_lock_period;
```
Comment (mark-tsyrulnyk) [inline contracts/contracts/rem.system/src/voting.cpp]: @AndriiAndrusyshyn as per @roman-tik remark, let this be; we will keep only `stake_lock_period` will have two constants: 1 for _lock period_ and one for _unlock period_
```
_voters.modify(vot, producer, [&](auto &v) {
    v.stake_lock_period = current_time_point() + stake_unlock_period;
}
```

File: contracts/contracts/rem.system/src/voting.cpp [hunk 8]
```diff
@@ -79,17 +79,28 @@ namespace eosiosystem {
             info.last_votepay_share_update = ct;
          });
          _gstate.total_producer_stake += tot.own_stake_amount;
+
+         const auto &vot = _voters.get(producer.value, "user has no resources");
```
Comment (mark-tsyrulnyk) [inline contracts/contracts/rem.system/src/voting.cpp]: @AndriiAndrusyshyn please move this code out of _else_ branch; it's should be applied for both new and existing producer

File: contracts/contracts/rem.system/src/delegate_bandwidth.cpp [hunk 9]
```diff
@@ -300,6 +300,13 @@ namespace eosiosystem {
           del_bandwidth_table     del_tbl( _self, from.value );
           const auto& del = del_tbl.get( from.value, "user has no resources" );
           check( current_time_point() > del.staked_time + stake_lock_period, "producer cannot undelegate bandwidth during 180 days");
```
Comment (mark-tsyrulnyk) [inline contracts/contracts/rem.system/src/delegate_bandwidth.cpp]: @AndriiAndrusyshyn please remove _stake lock_ logic from `del_bandwidth_table` we should keep it in `voter_info`

File: contracts/contracts/rem.system/src/rem.system.cpp [hunk 10]
```diff
@@ -19,7 +19,7 @@ namespace eosiosystem {
    const int64_t  default_annual_rate           = 0;       // 0% annual rate
    const int64_t  default_inflation_pay_factor  = 5;       // 20% of the inflation
    const int64_t  default_votepay_factor        = 4;       // 25% of the producer pay
-   const microseconds system_contract::vote_mature_period = eosio::days( 180 );
+   const microseconds system_contract::stake_lock_period = eosio::days( 180 );
```
Comment (mark-tsyrulnyk) [inline contracts/contracts/rem.system/src/rem.system.cpp]: @AndriiAndrusyshyn please add new _modifiable_  conf `stake_unlock_period` with the same default value

File: contracts/contracts/rem.system/src/delegate_bandwidth.cpp [hunk 11]
```diff
@@ -27,7 +27,7 @@ namespace eosiosystem {
    using std::pair;
 
    static constexpr uint32_t refund_delay_sec = 3*24*3600;
-   static const eosio::time_point stake_lock_period = eosio::time_point(eosio::days(180)); // two month
+   static const eosio::time_point stake_lock_period = eosio::time_point(eosio::days(180));
```
Comment (mark-tsyrulnyk) [inline contracts/contracts/rem.system/src/delegate_bandwidth.cpp]: @AndriiAndrusyshyn this const is outdated, please delete it

File: contracts/contracts/rem.system/src/delegate_bandwidth.cpp [hunk 12]
```diff
@@ -297,9 +296,13 @@ namespace eosiosystem {
 
       if(is_block_producer(from))
       {
-          del_bandwidth_table     del_tbl( _self, from.value );
-          const auto& del = del_tbl.get( from.value, "user has no resources" );
-          check( current_time_point() > del.staked_time + stake_lock_period, "producer cannot undelegate bandwidth during 180 days");
+          const auto &vot = _voters.get(from.value, "user has no resources");
+          _voters.modify(vot, from, [&](auto &v) {
+              check(v.locked_stake >= unstake_quantity.amount, "cannot undelegate more than was staked");
```
Comment (mark-tsyrulnyk) [inline contracts/contracts/rem.system/src/delegate_bandwidth.cpp]: @AndriiAndrusyshyn please move this check out of modification lambda and change error message in the second one to "cannot undelegate during stake lock period"

File: contracts/contracts/rem.system/src/delegate_bandwidth.cpp [hunk 13]
```diff
@@ -297,9 +294,13 @@ namespace eosiosystem {
 
       if(is_block_producer(from))
       {
-          del_bandwidth_table     del_tbl( _self, from.value );
-          const auto& del = del_tbl.get( from.value, "user has no resources" );
-          check( current_time_point() > del.staked_time + stake_lock_period, "producer cannot undelegate bandwidth during 180 days");
+          const auto &vot = _voters.get(from.value, "user has no resources");
+          check(vot.locked_stake >= unstake_quantity.amount, "cannot undelegate more than was staked");
+          _voters.modify(vot, from, [&](auto &v) {
+              check(v.stake_lock_time <= current_time_point(), "cannot undelegate during stake lock period");
```
Comment (mark-tsyrulnyk) [inline contracts/contracts/rem.system/src/delegate_bandwidth.cpp]: @AndriiAndrusyshyn this one too

**Review by mark-tsyrulnyk — APPROVED** [review]

**Review by me-vadim — APPROVED** [review]

**Pull request merged**